resteasy-client/src/main/java/org/jboss/resteasy/client/jaxrs/internal/ClientResponse.java

```diff
@@ -129,7 +130,24 @@ protected InputStream getEntityStream()
    {
       if (bufferedEntity != null) return new ByteArrayInputStream(bufferedEntity);
       if (isClosed()) throw new ProcessingException(Messages.MESSAGES.streamIsClosed());
-      return getInputStream();
+      InputStream is = getInputStream();
+      return is != null ? new InputStreamWrapper(is, this) : null;
+   }
+   
+   private static class InputStreamWrapper extends FilterInputStream {
+      
+      private ClientResponse response;
+      
+      protected InputStreamWrapper(InputStream in, ClientResponse response) {
+         super(in);
+         this.response = response;
+      }
+      
+      @Override
+      public void close() throws IOException {
+         super.close();
+         this.response.close();

```

Comment:
I could have simply set this.response = null here, but that would have been possibly "optimized" (removed) by an aggressive JIT compiler figuring out this.response wouldn't be used. So I'm actually explicitly closing the response if the inputstream is explicitly closed.